addons/portal/wizard/portal_wizard_user.py

```diff
@@ -0,0 +1,128 @@
+        error_empty = []
+        error_emails = []
+        error_user = []
+        ctx = dict(self.env.context or {}, active_test=False)
+        for wizard_user in self:
+            if wizard_user.in_portal and not wizard_user.user_id:
+                email = extract_email(wizard_user.email)
+                if not email:
+                    error_empty.append(wizard_user.partner_id)
+                elif email in emails and email not in error_emails:
+                    error_emails.append(wizard_user.partner_id)
+                user = res_users.sudo().with_context(ctx).search([('login', '=', email)])
+                if user:
+                    error_user.append(wizard_user.partner_id)
+                emails.append(email)
+
+        error_msg = []
+        if error_empty:
+            error_msg.append("%s\n- %s" % (_("Some contacts don't have a valid email: "),
+                                '\n- '.join(['%s' % (p.display_name,) for p in error_empty])))
+        if error_emails:
+            error_msg.append("%s\n- %s" % (_("Several contacts have the same email: "),
+                                '\n- '.join([p.email for p in error_emails])))
+        if error_user:
+            error_msg.append("%s\n- %s" % (_("Some contacts have the same email as an existing portal user:"),
+                                '\n- '.join(['%s <%s>' % (p.display_name, p.email) for p in error_user])))
+        if error_msg:
+            error_msg.append(_("To resolve this error, you can: \n"
+                "- Correct the emails of the relevant contacts\n"
+                "- Grant access only to contacts with unique emails"))
+        return error_msg
+
+    def _create_user(self):
+        """ create a new user for wizard_user.partner_id
+        """
+        values = {
+            'email': extract_email(self.email),
+            'login': extract_email(self.email),
+            'partner_id': self.partner_id.id,
+        }
+        return self.env['res.users'].with_context(noshortcut=True, no_reset_password=True).create(values)
+
+    @api.multi
+    def _send_email(self):
+        """ send notification email to a new portal user
+        """
+        self.ensure_one()
+        if not self.env.user.email:
+            raise UserError(_('You must have an email address in your User Preferences to send emails.'))
+
+        # determine subject and body in the portal user's language
+        context = dict(self.env.context or {}, lang=self.user_id.lang)
+        ctx_portal_url = dict(context, signup_force_type_in_url='')
+        partner = self.partner_id
+        portal_url = partner.with_context(ctx_portal_url)._get_signup_url_for_action()[partner.id]
+        partner.with_context(context).signup_prepare()
+        template = self.env.ref('portal.mail_template_data_portal_welcome')
+        if template:
+            template.with_context({'dbname': self.env.cr.dbname, 'portal_url': portal_url}).send_mail(self.id, force_send=True)
```

Comment:
semantics change. Old code added keys to existing context, this one completely replaces context to only use these keys.